Get institution routes

yunkaa19 opened the pull request:
The get institution controler/presenter/usecase have been completed.

yunkaa19 committed:
Get institution routes

The get institution controler/presenter/usecase have been completed.

FelipeCarillo commented on src/modules/get_institution/app/get_institution_presenter.ts:
```diff
@@ -0,0 +1,16 @@
+const usecase = new GetInstitutionUsecase(repository.InstitutionRepo);
+const controller = new GetInstitutionController(usecase);
+
+export const handler = async (event: any, context: any) => {
```
Check the others presenters! U need to change the function handler!

FelipeCarillo commented on src/modules/get_institution/app/get_institution_usecase.ts:
```diff
@@ -0,0 +1,21 @@
+        this.database_repo = database_repo;
+    }
+
+    public async execute(id: string) {
```
U need to make all the logic of getting the authorization, check if the user is register... U can check in the other routes, how to do it!

FelipeCarillo commented on src/modules/get_institution/app/get_institution_controler.ts:
```diff
@@ -0,0 +1,45 @@
+    public async execute( id: string): Promise<{ [key: string]: any }> {
+        try {
+            
+            
```
The same case! Check in the other routes, what controller receive, what is it logic...

If you have some questions or difficulties, send a message to me!

yunkaa19 committed:
Updated accorting to Felipe Requirements

FelipeCarillo commented on src/modules/get_institution/app/get_institution_usecase.ts:
```diff
@@ -0,0 +1,43 @@
+        .catch((error) => {
+            throw new UserNotAuthenticated();
+        });
+        
```
You need verify if the user is registered in our database.

FelipeCarillo commented on src/modules/get_institution/app/get_institution_usecase.ts:
```diff
@@ -0,0 +1,43 @@
+    public token_auth: TokenAuth;
+    public institution_repo: IInstitutionRepo;
+
+    constructor(institution_repo: IInstitutionRepo) {
```
Add user_repo.

FelipeCarillo commented on src/modules/get_institution/app/get_institution_presenter.ts:
```diff
@@ -0,0 +1,16 @@
+import { Repository } from "../../../core/repositories/Repository";
+import { HttpRequest } from "../../../core/helpers/http/http_codes";
+
+const repository = new Repository({institution_repo: true});
```
Need to add true value to user_repo.

FelipeCarillo commented on src/modules/get_institution/app/get_institution_presenter.ts:
```diff
@@ -0,0 +1,16 @@
+
+const repository = new Repository({institution_repo: true});
+
+const usecase = new GetInstitutionUsecase(repository.InstitutionRepo);
```
Need to add the repository.UserRepo

FelipeCarillo commented on src/modules/get_institution/app/get_institution_presenter.ts:
```diff
@@ -0,0 +1,16 @@
+
+export const handler = async (event: any, context: any) => {
+    let request = new HttpRequest(event);
+    let response = await controller.executeexecute(request);
```
Fix the double execute.

FelipeCarillo commented on src/modules/get_institution/app/get_institution_usecase.ts:
```diff
@@ -0,0 +1,43 @@
+        this.token_auth = new TokenAuth();
+    }
+
+    async execute(headers: { [key: string]: any}) {
```
Add queryStringParameters

FelipeCarillo commented on src/modules/get_institution/app/get_institution_controler.ts:
```diff
@@ -0,0 +1,54 @@
+                {
+                    throw new InvalidRequest("Headers");
+
+                }
+
+            let response = await this.usecase.execute(request.headers);
```
Verify if the request.body exist
Them pass to the usecase

yunkaa19 committed:
Updated to the requests of Felipe!

FelipeCarillo commented on src/modules/get_institution/app/get_institution_usecase.ts:
```diff
@@ -0,0 +1,57 @@
+            throw new UserNotAuthenticated();
+        }
+
+        if (user.user_type === UserTypeEnum.STUDENT) {
```
change this condition to:
```
if (!([UserTypeEnum.MODERATOR, UserTypeEnum.ADMIN].includes(user.user_type))
```
This condition is better because it avoid that user with different types can access this route.

FelipeCarillo commented on src/modules/get_institution/app/get_institution_usecase.ts:
```diff
@@ -0,0 +1,57 @@
+            {
+                throw new MissingParameter("Authorization");
+            }
+
```
Verify if institutionData and institutionData.institution_id isn't null.

FelipeCarillo commented on src/modules/get_institution/app/get_institution_presenter.ts:
```diff
@@ -0,0 +1,18 @@
+import { Repository } from "../../../core/repositories/Repository";
+import { HttpRequest } from "../../../core/helpers/http/http_codes";
+
+const InstituteRepository = new Repository({institution_repo: true});
+const UserRepository = new Repository({user_repo: true});
```
You instance the const Repository and pass:
```
const Repository = new Repositoy({ user_repo: true, institution_repo: true })
``` 

is no need to create two Repos!

FelipeCarillo commented on src/modules/get_institution/app/get_institution_presenter.ts:
```diff
@@ -0,0 +1,18 @@
+const UserRepository = new Repository({user_repo: true});
+
+
+const usecase = new GetInstitutionUsecase(InstituteRepository.InstitutionRepo, UserRepository.UserRepo);
```
change to Repository.InstitutionRepo and Repository.UserRepo

FelipeCarillo commented on src/modules/get_institution/app/get_institution_controler.ts:
```diff
@@ -0,0 +1,57 @@
+                throw new InvalidRequest("Body");
+            }
+
+            let response = await this.usecase.execute(request.body.body ,request.headers);
```
here is the queryStringParameters, because is a get method so pass to usecase the request.body.queryStringParameters

yunkaa19 committed:
Resolved the pull request changes

FelipeCarillo commented on src/modules/get_institution/app/get_institution_controler.ts:
```diff
@@ -0,0 +1,57 @@
+            }
+            if (error instanceof UserNotAuthenticated) {
+                return new Unauthorized(error.message);
+            }
```
Put a if here, if is UserNotAllowed, return a Forbidden!

FelipeCarillo reviewed (changes requested):
Make the unit test for presenter only!

yunkaa19 committed:
testing phase

yunkaa19 committed:
Update get_instition.test.ts

FelipeCarillo committed:
refactor: Add new error classes for user authentication and not found errors

FelipeCarillo committed:
feat: Add description field to InstitutionMock

The `InstitutionMock` class now includes a `description` field to provide a description for each institution.

FelipeCarillo committed:
chore: Update get_institution_controller.ts

Refactor the import statements in `get_institution_controller.ts` to include the new error classes for user authentication and not found errors.

FelipeCarillo committed:
refactor: Update get_institution_usecase.ts

Refactor the `get_institution_usecase.ts` file to improve code readability and remove unnecessary imports and conditions.

FelipeCarillo committed:
feat: Refactor Get Institution Presenter

Refactor the Get Institution Presenter to improve code readability and remove unnecessary imports and conditions.
